### PR TITLE
Checkpoint resuming compatible for 2403 container

### PR DESCRIPTION
**Pull request by suiyoubi**
# What does this PR do ?

Fix the issue that ckpt trained in 2403 cannot be resumed in latest NeMo

**Collection**: NLP

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [X] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 1]
```diff
@@ -439,7 +440,31 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
             # after dist_checkpointing.load, sharded tensors will be replaced with tensors
             checkpoint['state_dict'] = sharded_state_dict
             checkpoint['optimizer_states'] = [self.optimizer_sharded_state_dict()]
-            return self.checkpoint_io.load_checkpoint(checkpoint_path, sharded_state_dict=checkpoint)
+
```
Comment (akoumpa) [inline nemo/collections/nlp/parts/nlp_overrides.py]: can you make this whole block another function? also can you add a docstrings what the (new function) is doing?

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 2]
```diff
@@ -439,7 +440,31 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
...
+                        # Temporary empty params so that loading doesn't fail
+                        model_param_groups.insert(expert_index, {'params': LocalNonpersitentObject([]), 'is_expert': True})
+                        checkpoint['optimizer_states'][0]['param_groups'] = model_param_groups
+                    else:
+                        raise ValueError('Cannot find expert param in the checkpoint.')
+                    
+            loaded_state_dict = self.checkpoint_io.load_checkpoint(checkpoint_path, sharded_state_dict=checkpoint)
+            if expert_index is not None:
+                # Remove the temporary empty params added above
+                loaded_state_dict['optimizer_states'][0]['param_groups'].pop(expert_index)
```
Comment (akoumpa) [inline nemo/collections/nlp/parts/nlp_overrides.py:477]: can you add an assert statement here that ensures that param_groups of expert_index is empty?

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 3]
```diff
@@ -439,7 +440,31 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
...
             checkpoint['state_dict'] = sharded_state_dict
             checkpoint['optimizer_states'] = [self.optimizer_sharded_state_dict()]
-            return self.checkpoint_io.load_checkpoint(checkpoint_path, sharded_state_dict=checkpoint)
+            
+            common_state_dict = dist_checkpointing.load_common_state_dict(checkpoint_path)
+            checkpoint_param_groups = common_state_dict['optimizer_states'][0]['param_groups']
+            model_param_groups = checkpoint['optimizer_states'][0]['param_groups']
+            expert_index = None
+            if len(checkpoint_param_groups) != len(model_param_groups):
+                # In 24.03, all checkpoints store EP param regardless of using EP or not
```
Comment (akoumpa) [inline nemo/collections/nlp/parts/nlp_overrides.py:449]: "EP param" -> "EP param group"

**Review by akoumpa — APPROVED** [review]
thank you so much for the refactor.
